sw/device/tests/sram_ctrl_execution_test_ret.c

```diff
@@ -0,0 +1,175 @@
+ *
+ * This exception handler keeps default exception implementations for faults
+ * that are not relevant in terms of this test. The only fault handler that
+ * is overridden is Instruction Access Fault.
+ *
+ * The controlled Instruction Access Fault occurs originates in the retention
+ * SRAM, which means that normally the return address would be calculated
+ * relevant to the offending instruction. However, due to execution from
+ * retention SRAM being permanently disabled, this approach would not work.
+ * The return address for the one stack frame before is used instead. In
+ * other words sram_execution_test -> retention_sram -> exception_handler
+ * -> sram_execution_test.
+ *
+ * Before the jump into the exception handler, the register set is saved on
+ * stack. This means that the return address can be loaded from the stack, or
+ * simply read from the return address register (before it is overridden by
+ * the code in this exception handler). This means that the return address
+ * register value must be read before any function call.
+ */
+void handler_exception(void) {
+  // Save the return address register value.
+  uint32_t ret_addr = 0;
+  asm volatile("sw ra, %0" : "=m"(ret_addr));
+
+  uint32_t mcause;
+  exc_id_t exc_cause;
+
+  CSR_READ(CSR_REG_MCAUSE, &mcause);
+  exc_cause = (exc_id_t)(mcause & kIdMax);
+
+  switch (exc_cause) {
+    case kInstMisa:
+      handler_instr_acc_fault();
+      break;
+    case kInstAccFault:
+      LOG_INFO("Instruction access fault handler");
+      instruction_access_fault = true;
+      fault_handler_testutils_set_return_addr(&ret_addr);
+      break;
+    case kInstIllegalFault:
+      handler_instr_ill_fault();
+      break;
+    case kBkpt:
+      handler_bkpt();
+      break;
+    case kLoadAccFault:
+      handler_lsu_fault();
+      break;
+    case kStrAccFault:
+      handler_lsu_fault();
+      break;
+    case kECall:
+      handler_ecall();
+      break;
+    default:
+      while (1) {
+      };
+  }
+}
+
+/*
+ * Performs SRAM execution test.
+ *
+ * `ram_exec_function_t` is "mapped" onto the `execution_test_instructions`
+ * buffer that holds hand-crafted machine instructions. Effectively, this
+ * buffer becomes the function body.
+ *
+ * According to the calling convention the parameters will be passed in the
+ * a0 and a1 registers, with a0 also treated as the return value register.
+ */
+static void sram_execution_test(void) {
+  memcpy((void *)kRamStartAddr, &execution_test_instructions,
+         sizeof(execution_test_instructions));
+
+  // Map the function pointer onto the instruction buffer.
+  sram_ctrl_testutils_exec_function_t func =
+      (sram_ctrl_testutils_exec_function_t)kRamStartAddr;
+
+  uintptr_t func_address = (uintptr_t)func;
+  CHECK(func_address >= kRamStartAddr && func_address <= kRamEndAddr,
+        "Test code resides outside of the Ret SRAM: function address = %x",
+        func_address);
+
+  instruction_access_fault = false;
```

Comment:
This is a global so you don't need to manually initialize it.

Reply:
It doesn't harm, I think I will leave it.

Reply:
Can you please remove it? It's not necessary.

Reply:
I disagree with removing it. Setting the flag to false explicitly here helps with code readability as it is clearer when skimming the code what the value should be if there is no exception.

Reply:
Keeping the flag, but I should have explained the intention better (@mundaym captured how I feel about it very well in his comment).